onnxruntime/core/common/threadpool.cc

```diff
@@ -378,5 +376,52 @@ int ThreadPool::CurrentThreadId() const {
   }
 }
 
+void ThreadPool::TryParallelFor(concurrency::ThreadPool* tp, std::ptrdiff_t total, const TensorOpCost& cost_per_unit,
+                           const std::function<void(std::ptrdiff_t first, std::ptrdiff_t last)>& fn) {
+#ifdef _OPENMP
+    if (total == 1) {
+      fn(0, total);
+      return;
+    }
+
+    Eigen::TensorOpCost cost{cost_per_unit.bytes_loaded, cost_per_unit.bytes_stored, cost_per_unit.compute_cycles};
+    auto d_of_p = DegreeOfParallelism(tp);
+    std::ptrdiff_t num_threads = CostModel::numThreads(static_cast<double>(total), cost, d_of_p);
+
+    if (total < num_threads) {
+      num_threads = total;
+    }
+
+    if (num_threads == 1) {
+      fn(0, total);
+      return;
+    }
+
+    ptrdiff_t block_size = CalculateParallelForBlock(total, cost, nullptr, d_of_p);
+    ptrdiff_t block_count = Eigen::divup(total, block_size);
+    if (block_count < num_threads) {
+      num_threads = block_count;
+    }
+
+    if (num_threads == 1) {
+      fn(0, total);
+      return;
+    }
+
+#pragma omp parallel for
+    for (std::ptrdiff_t i = 0; i < num_threads; i++) {
+      auto work = PartitionWork(i, num_threads, total);
+      fn(work.start, work.end);
+    }
```

Comment:
This loop won't match the behavior of the Eigen variant.  With Eigen, we pick the number of threads to use (usually all of them, unless there are very few iterations [which may well be the case in the tests with regressions]).  We then pick the number of iterations to give to each thread whenever it claims a chunk of work (say, 1024 iterations from a total of 100000).  The loop then has threads dynamically claim new chunks of work whenever they finish their current work.  The OpenMP variant here will pick the number of threads, but then statically partition the work between the threads, rather than using the block_size that was computed.

The closest OpenMP variant would be

#pragma omp parallel for schedule(dynamic,1)
for (auto i = 0; i < block_count; i++) {
     const auto start = i*block_size;
     fn(start, ::std::min(start+block_size, total));
}

Reply:
Can we do it without calling PartitionWork?
```c++
for (std::ptrdiff_t i = 0; i < total; i++) {
  fn(i);   //we need a different signature of fn for such purpose.
}
```

Reply:
@snnn i guess your proposal doesn't make use *block_size*.

Reply:
how about the following?
#pragma omp parallel for schedule(dynamic, block_size)
for (std::ptrdiff_t i = 0; i < total; i++) {
  fn(i);   //we need a different signature of fn for such purpose.
}

Reply:
I wonder if openmp can calculate the block size better than us ?

Reply:
Good question, that's something I'm not sure.

Reply:
@snnn @duli2012 OpenMP does not do anything smart like analyzing the code to estimate costs.  It just distributes the work between threads based on the schedule(...) directive.  In principle we could use schedule(dynamic, block_size), but the advantage of schedule(dynamic,1) and then calculating bounds inside the loop is that:

- We use the same two-parameter function in the other versions of the API

- We have one call in/out of fn() for each block of iterations.  This lets the compiler optimize within each of those blocks; that can be significant for loops with small amounts of work in each iteration.  [To achieve this with schedule(dynamic,block_size) we would probably need to pull that loop up into the header file, and then ensure inlining of the body of fn() into the loop.  I have done that before via C++ templates, but it's probably adding a code size overhead for little perf benefit beyond the low-tech two-parameter approach we have here.]